### PR TITLE
add second palette

### PR DESCRIPTION
**Pull request by spacek531**
Custom content creators struggle with handling 8-bit png when creating and editing sprites, and most tools to create custom content for RCT2 utilise a different palette than OpenRCT2. The OpenRCT2 palette does not have unique RGB values for every palette color - several animated water colors have the same RGB value, which cause problems when importing and exporting in true color.

This PR attempts to address both issues by providing native support for the RCT2 palette as well defining one's own palette to import/export with. 

todo:

- [x] plan of action for animated water colors

features:

- [x] color 0 exports as opaque on custom palettes with opaque 0th pixel

testing plan:

- build gx using 3 different palettes
- append image to gx using 3 different palettes
- export gx using 3 different palettes
- exportall using 3 different palettes
- exportalldat using 3 different palettes
- build g2 and run game
- run game with parkobj with image loading (X123M3-256's trains work for this)

testing results:

- [x] build with green
- [x] build with custom
- [x] append with green
- [x] append with custom
- [x] export with green
- [x] export with custom
- [x] exportall with green
- [x] exportall with custom
- [x] exportalldat with green
- [x] exportalldat with custom

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
Wouldn't the better option be that you can specify a user defined palate? What happens when someone wants to use it with some other remap colour.

**Comment by spacek531** [comment]
I did think of that, but I didn't think of an elegant way to add it to the sprite entry of a sprite list manifest.

**Comment by duncanspumpkin** [comment]
You could just provide a 1x256 bitmap of your colours. I think that would work.

**Comment by spacek531** [comment]
that is a good idea

**Comment by ZehMatt** [comment]
I have to agree with @duncanspumpkin I don't think this is practical doing it this way, user defined palettes would avoid adding code for every specific case.

**Comment by IntelOrca** [comment]
We also have an option to use the raw indices that are stored in .png file. This is faster and would allow you to use any colour, because the game would only use the raw byte index for each pixel rather than using a colour lookup table.

**Comment by spacek531** [comment]
I know that, but the tools to create 8-bit images elude me, and this was the easiest option.

**Comment by Gymnasiast** [comment]
> Wouldn't the better option be that you can specify a user defined palate? What happens when someone wants to use it with some other remap colour.

I should point out that every other tool in existence uses green as the primary remap instead of orange. As such, I think building it in makes more sense in this case.

**Comment by duncanspumpkin** [comment]
> > Wouldn't the better option be that you can specify a user defined palate? What happens when someone wants to use it with some other remap colour.
> 
> I should point out that every other tool in existence uses green as the primary remap instead of orange. As such, I think building it in makes more sense in this case.

As that is not true that is the reason why i suggested having the ability to provide any palette.

**Comment by spacek531** [comment]
> > As that is not true that is the reason why i suggested having the ability to provide any palette.
> 
> Ok, can you name a tool that doesn’t use green for primary remap? The Object Editor, Trigger’s tools and the RCT2 GUI Editor all use that palette.

Landmark editor uses an alternative palette. I vaguely recall one other, but I don't remember its name or what it did.

**Comment by spacek531** [comment]
I'm open to suggestions on making the command line parsing less bad, as well as setting the actual data of each PaletteBGRA

**Comment by Gymnasiast** [comment]
> > > As that is not true that is the reason why i suggested having the ability to provide any palette.
> > 
> > Ok, can you name a tool that doesn’t use green for primary remap? The Object Editor, Trigger’s tools and the RCT2 GUI Editor all use that palette.
> 
> Landmark editor uses an alternative palette. I vaguely recall one other, but I don't remember its name or what it did.

Do you have a link? I tried Googling, but didn’t see anything that looked like a RCT sprite editor/exporter.

**Comment by spacek531** [comment]
> Do you have a link? I tried Googling, but didn’t see anything that looked like a RCT sprite editor/exporter.

https://www.nedesigns.com/index.php?app=core&module=attach&section=attach&attach_id=25880

From the NEDesigns website. IIRC there are two links, one of them worked when I tried it and one of them didn't. I don't remember where the other link is.

**Comment by spacek531** [comment]
I found an issue with the example palettes, the 0th pixel is 255 255 255 255 instead of 0 0 0 0. I'm not in a position to fix it right now, so if there's any danger of this PR being merged in the next several hours, please hold back until they are fixed.

**Comment by spacek531** [comment]
> I found an issue with the example palettes, the 0th pixel is 255 255 255 255 instead of 0 0 0 0. I'm not in a position to fix it right now, so if there's any danger of this PR being merged in the next several hours, please hold back until they are fixed.

This issue has been resolved. I think the PR is ready for review, less any suggestions regarding what I asked about earlier.

**Comment by spacek531** [comment]
I have been thinking about OpenRCT2's palette collisions with sparkling water and I wanted to see if any tools do not have palette collisions. It appears none of the tools avoid palette collisions, I think because no scenery actually uses all of the animated water colors.

I've written more of my findings here: https://github.com/OpenRCT2/OpenRCT2/discussions/16611

As far as the effect on this PR, since there is no standard on which to base these colors on, I think this PR is an opportunity to create a new standard for animated water pixels, unless it is preferred to update these colors in another branch.

File: src/openrct2/drawing/ImageImporter.h
Comment (duncanspumpkin) [inline src/openrct2/drawing/ImageImporter.h:77]: You are passing this function two palettes that makes no sense. It should only take the new palette (const ref). You will need to modify the caller's a bit to handle this

File: src/openrct2/drawing/ImageImporter.h
Comment (duncanspumpkin) [inline src/openrct2/drawing/ImageImporter.h:60]: As below two palettes

File: src/openrct2/CmdlineSprite.cpp
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:244]: This looks wrong. You should be reading the png directly into the palette.

Comment (spacek531) [reply src/openrct2/CmdlineSprite.cpp:244]: This code is primarily copied from the image import code, so I don't know what you are looking for in it.

File: src/openrct2/object/ImageTable.cpp
Comment (duncanspumpkin) [inline src/openrct2/object/ImageTable.cpp:178]: why are you copying the palette here?

File: src/openrct2/core/Imaging.cpp
Comment (duncanspumpkin) [inline src/openrct2/core/Imaging.cpp:224]: why this change?

File: src/openrct2/CmdlineSprite.cpp
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:706]: I don't see why this is being tied to the palette they are unrelated options.

File: src/openrct2/CmdlineSprite.cpp
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:204]: the image is height*width we don't need to mess about with any of this you can just do `for (int i = 0; i < 256; ++i)` stride doesn't matter.

File: src/openrct2/core/Imaging.cpp
Comment (spacek531) [reply src/openrct2/core/Imaging.cpp:224]: Custom palettes can choose an opaque or transparent background

File: src/openrct2/object/ImageTable.cpp
Comment (spacek531) [reply src/openrct2/object/ImageTable.cpp:178]: Aborted change to this function to use custom palettes, I will do that in another PR, I forgot to remove this line.

File: src/openrct2/core/Imaging.cpp
Comment (duncanspumpkin) [reply src/openrct2/core/Imaging.cpp:224]: What relevance does that have to creating a png?

Comment (spacek531) [reply src/openrct2/core/Imaging.cpp:224]: If the user wants an opaque background color for their exported png, they can get an opaque background color by setting the first pixel of the palette to opaque

**Comment by spacek531** [comment]
It's been a week since the last action and this PR is now in a holding pattern. I've tested all the functionality and I am happy with it. Is there anything else left before this PR can get merged?

File: src/openrct2/drawing/ImageImporter.h
Comment (duncanspumpkin) [inline src/openrct2/drawing/ImageImporter.h:69]: pass by const ref.

File: src/openrct2/CmdlineSprite.cpp
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:215]: Should really be using a string_view for this.

File: src/openrct2/CmdlineSprite.cpp
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:187]: paths should be utf8* not char*

File: src/openrct2/CmdlineSprite.cpp
Comment (duncanspumpkin) [inline src/openrct2/CmdlineSprite.cpp:652]: you use a case insensitive compare in `GetPaletteFromString` but a case sensitive compare here. Stick to one. (probably just the case sensitive one)

**Review by duncanspumpkin — APPROVED** [review]
i think this will be fine. Few minor corrections to conform more with the codebase.

File: src/openrct2/CmdlineSprite.cpp
Comment (oli414) [inline src/openrct2/CmdlineSprite.cpp:224]: This doesn't seem safe cross-platform, aren't there some path related helper functions?

Comment (Broxzier) [reply src/openrct2/CmdlineSprite.cpp:224]: C++ has [`std::filesystem::path`](https://en.cppreference.com/w/cpp/filesystem/path). Include `core/FileSystem.hpp` for this.

Comment (duncanspumpkin) [reply src/openrct2/CmdlineSprite.cpp:224]: None of the cmdline code is safe cross platform.

**Comment by spacek531** [comment]
I do not have a plugin that uses custom images to test against.

**Comment by IntelOrca** [comment]
@spacek531

* https://github.com/OpenRCT2/title-sequence-editor/releases/tag/v1.2
* https://github.com/OpenRCT2/OpenRCT2/files/8346277/image-api-demo.zip

**Comment by spacek531** [comment]
The image api demo works fine in this branch. Please can this be merged before any other changes are made?

**Comment by tupaschoal** [comment]
> The image api demo works fine in this branch. Please can this be merged before any other changes are made?

Before which other changes?

**Comment by spacek531** [comment]
Any others that affect image import code that I have to work around and require another review and delay things further until the next change that affects image import code that I have to work around and require another review that delay things further and so on.

**Comment by spacek531** [comment]
I've tested all the functionality and I am happy with it. Is there anything else left before this PR can get merged?

**Comment by spacek531** [comment]
Is this PR now defunct since libsawyer is being created? I wish there was better communication.

**Comment by tupaschoal** [comment]
@IntelOrca ?

**Comment by IntelOrca** [comment]
These changes can be ported now to gxc in libsawyer.
I guess there is an argument for this still existing in object and plugin, but it looks like we are heading more into embedding GX files now for best performance.

**Comment by spacek531** [comment]
It feels like I'm missing something critical in libsawer... I've poked around but I have yet to find where the sprite functions are actually used.

**Comment by IntelOrca** [comment]
> It feels like I'm missing something critical in libsawer... I've poked around but I have yet to find where the sprite functions are actually used.

I am not sure what you are looking for.

**Comment by karst** [comment]
Here's two color table files for the palette index. OpenRCT2 is default gxc, OpenRCT2 green is the green palette used here. You can easily convert between the two by just swapping the index.

[Palette color tables](https://github.com/OpenRCT2/OpenRCT2/files/10272662/Palette.color.tables.zip)

**Comment by Gymnasiast** [comment]
Keep this open please.

**Comment by Gymnasiast** [comment]
Keep this open please.

**Pull request closed without merging**
